README
==============================================
**serial functional Diffusion Mapping (sfDM)**
==============================================

Reference in Cancer Informatics:

https://www.ncbi.nlm.nih.gov/pubmed/25673970


INSTALLATION
============

Dependencies
------------
The following software must be installed first:

- fsl
- afni
- python-nipy
- python-nipype

The following will be installed along with the sfDM package if not already present:

- numpy
- scipy
- matplotlib
- pillow
- networkx
- ipython

Installing
----------

From Pypi:
pip install sfDM

From github:

git pull https://github.com/PIRCImagingTools/sfDM

cd into directory

python setup.py install


Running
=======

Following installation, type ``sfDM_gui`` in a terminal to lauch the gui

Pre-processing
--------------
Tumor region of interest for each time point must be manually drawn and saved as a binary mask in nifti format (must be in the same 3D space as the DWI images)

High resolution reference image must be brain extracted first. A high quality, properly brain-extracted image here can make the difference between successful registration and failing. This step is not automated to allow the option of different image contrasts/protocols to serve as the High Res template.

Data set up
-----------
Previously generated configuration file can be loaded with the *Load Config* Button. Make sure to save the configuration once the information is entered before running the first step.

Patient Info
------------

1) **PID**: Patient ID to be used in file naming and identification.

2) **Parent Dir**: Directory where output files (as well as crash logs) will be saved.

3) **Struct Brain**: Location of brain extracted High Res reference image.

4) **Tumor Center**: *X,Y,Z* coordinates of the center of the tumor (in High Res reference space). This will be used to generate 3-plane images of each FDM map.

5) **Treatment**: If the patient began any treatment of interest, check to create it on the fdm timeline. Days post baseline is the number of days after the baseline scan (first time point)

6) **Use GridEngine**: Check to run the registrations in a parallel environment (*experimental*). The environment must be properly configured. The registrations will be submitted to long.q

7) **Nonlinear Reg**: This step will use FSL's FNIRT non-linear registration to register each time point into reference space.(*We are currently optimizing the registration parameters*).

8) **Reg ADC**: In the event that the original B0 images are not available (only ADCs), the ADC maps can be used directly to register to the reference images. While B0's are preferable, this method has yielded good results. 

Scan Info
---------
Use the +/- Scan buttons to add or remove time points

1) **DWI**: Location of B0 image (nii/nii.gz). This can be the stand-alone B0, or 4D stacked B0/S1. If the *Reg ADC* option is checked, this will not be used and may be left blank

2) **ADC**-*Required*: Location of ADC image (nii/nii.gz). If registering to B0 image, they must both be in the same 3D space.

3) **REF**-*Required*: Location of medium to high resolution image, preferably T2 weighted acquired at the same time to serve as intermediate registration image (nii/nii.gz).

4) **ROI**-*Required*: Location of manually delineated tumor ROI in the same 3D space as the ADC/B0 images (nii/nii.gz).

5) **Time(Days)**-*Required for Time line*: Time in days from baseline scan of the current time point.

**Remember to save any configuration changes before running!**

Reg
___
This will register every time point into common space using a series of linear registrations (and a final non-linear reg is the option is selected).

Keep an eye on the terminal for any errors and warnings. 

Upon completion, use the **View** button to inspect the registrations. This will load the *FDM/Outputs/MergedRegs.nii.gz* file into fslview for convenient inspecting. This is a 4D stack of all the registrations, overlaid over the *Struct Brain* image. 

**Important**: Ensure that all registrations are visibly accurate before proceeding, or else the analysis will be invalid. 

Time Points
___________

Here you select from the list the *fDM Maps* to be generated. There are two lists: Baseline and Serial. By convention, baseline should be all time points compared back to the baseline scan, while serial is every subsequent scan compared to the previous.

You may, however, also include any additional comparisons you wish to make. 

*Only the time points added to the lists will have fDM Maps generated*

**Make sure to SAVE the time line when done**

Maps
____

This will generate the maps,plots, and images chosen in the previous step, as well as output fDM metrics to the files *serial.csv* and *baseline.csv* in the *parent_dir*

All generated images are saved in *FDM/Outputs* in the parent_dir

Use the **View** Button to load the generated maps into FSL

Time Line
_________

If desired, this step will generate and output the the patient's fDM timeline to the parent_dir.







CHANGES
Updated gui runner
Modified AFNI registration

0.3
---
Added option for Ref space tumor masks
New splash screen!

0.3.2
---
Added option to specify own threshold 


sfDM/fdm/fdm_timeline.py
"""
Draw time line from list of images
Required:
    timeline.json files listing the files in the desired order
"""

from PIL import Image,ImageDraw, ImageFont
import json,os
import numpy as np


config_file = os.environ['fdm_config']
timeline_file = os.environ['fdm_timeline']


with open(config_file, 'r') as f:
    cfg = json.load(f)

with open(timeline_file, 'r') as f:
    timeline = json.load(f)

PID=cfg['PID']

BaseDir = cfg['parent_dir']
ImageDir = BaseDir+'/FDM/Outputs/Images/'
PlotDir = BaseDir+'/FDM/Outputs/Plots/'
MetricDir = BaseDir+'/FDM/Outputs/Metrics/'

Treatment = cfg['treatment']



##################################################################################

print "Creating Time Line"

baseline_scan_list = timeline['baseline']
serial_scan_list = timeline['serial']

baseline_list = ['scan_{0}-scan_{1}.png'.format(scan[5:7],scan[15:17]) for
                                       scan in baseline_scan_list]
serial_list = ['scan_{0}-scan_{1}.png'.format(scan[5:7],scan[15:17]) for
                                       scan in serial_scan_list]


baseline_time = [int(cfg[scan[:7]]['time']) for
                                        scan in baseline_scan_list]
serial_time = [int(cfg[scan[:7]]['time']) for
                                        scan in serial_scan_list]
if Treatment:
    basedate = int(cfg['v_days'])
    scans_before = np.size(np.where(np.array(baseline_time) < basedate))
else:
    scans_before = 0

p_width = 533
p_height = 400
m_width = 533
m_height = 400
vspace = 30
time_unit = 5 #pixels


basedate= int(basedate) * time_unit + p_width*(scans_before+1)
longest = np.max(np.vstack((serial_time, baseline_time)))

image_width=len(baseline_list)*(p_width) +int(longest)*time_unit + m_width+ 10
image_height=(p_height*4 + vspace*5) + 10

baseline_plot_files=[PlotDir+plot for plot in baseline_list]
baseline_plot_images=[Image.open(filename) for filename in baseline_plot_files]

baseline_map_files=[ImageDir+dmap for dmap in baseline_list]
baseline_map_images=[Image.open(filename) for filename in baseline_map_files]

serial_plot_files=[PlotDir+plot for plot in serial_list]
serial_plot_images=[Image.open(filename) for filename in serial_plot_files]

serial_map_files=[ImageDir+dmap for dmap in serial_list]
serial_map_images=[Image.open(filename) for filename in serial_map_files]

# create metrics file
baseline_metrics = [MetricDir+csv[:-3]+'csv' for csv in baseline_list]
serial_metrics = [MetricDir+csv[:-3]+ 'csv' for csv in serial_list]

def write_metrics(filelist, outfile):
    csvout = open(outfile, 'wb')
    with open(filelist[0], 'r') as f:
            csvout.write(f.read())
    for csv in filelist[1:]:
        with open(csv, 'r') as f:
            f.next()
            for line in f:
                csvout.write(line)
    csvout.close()

write_metrics(baseline_metrics, BaseDir+'/baseline.csv')
write_metrics(serial_metrics, BaseDir+'/serial.csv')


img = Image.new("RGB",(image_width,image_height), (255,255,255))   #w, h

#Lines

draw = ImageDraw.Draw(img)
draw.rectangle((0, p_height/2 + vspace+50,    #### First black line
            image_width,p_height/2 + vspace + 55),
            fill='#000000')  #x1,y1, x2,y2
draw.rectangle((0, 2.5*p_height + 3*vspace,     #### Second black line
           image_width, 2.5*p_height + 3*vspace + 5),
           fill='#000000')  #x1,y1, x2,y2
draw.rectangle((0, image_height/2 - 85,        #### Middle blue time line
           image_width,image_height/2 - 70),
           fill='#0000ff')  #x1,y1, x2,y2
if Treatment:
    draw.rectangle((basedate-5, 0,                  ##### Vertical baseline
           basedate-1,image_height),
           fill='#ff0000')  #x1,y1, x2,y2


    #WEIGHTS
weight_image = Image.open(ImageDir+PID+'_Weights.png')
w,h = weight_image.size
weight_image = weight_image.crop((int(w/2.5) , int(h/15), w - int(w/3), h - int(h/20) ))
WeightMap = weight_image.resize((int(p_width*1.2), p_height*3), Image.ANTIALIAS)
ImageDraw.floodfill(WeightMap, (0,0), (255,255,255))
img.paste(WeightMap, (0, vspace + 70))


fontpath='/usr/share/fonts/truetype/ttf-dejavu/DejaVuSans.ttf'
font = ImageFont.truetype(fontpath, 60)
draw.text((p_width, image_height/2-70), 'Days from first scan', fill='#0000ff', font=font)

if Treatment:
    draw.text((basedate+15,3), 'Baseline treatment', fill='#ff0000', font=font)


for i in range(len(baseline_map_images)):
    BasePlot = baseline_plot_images[i].resize((p_width,p_height), Image.ANTIALIAS)
    BaseMap = baseline_map_images[i].resize((m_width, m_height), Image.ANTIALIAS)
    ImageDraw.floodfill(BaseMap,(0,0),(255,255,255))

    SerialPlot = serial_plot_images[i].resize((p_width, p_height), Image.ANTIALIAS)
    SerialMap = serial_map_images[i].resize((m_width, m_height), Image.ANTIALIAS)
    ImageDraw.floodfill(SerialMap,(0,0), (255,255,255))


    img.paste(BaseMap, (time_unit*int(baseline_time[i]) + p_width*(i+1),
                        p_height+1*vspace-70))
    img.paste(SerialMap, (time_unit*int(serial_time[i])+ p_width*(i+1),
                         3*p_height+3*vspace-70))

    img.paste(BasePlot, (time_unit*int(baseline_time[i]) + p_width*(i+1),
                        vspace+50))
    img.paste(SerialPlot, (time_unit*int(serial_time[i])+ p_width*(i+1),
                        2*p_height+3*vspace))

    draw.text((time_unit*int(baseline_time[i]) + p_width*(i+1) + p_width/2 -20,
                image_height/2 - 160),
                str(baseline_time[i]), fill='#0000ff', font=font)



img.save(BaseDir+'/'+PID+'_w-timeline.png')


sfDM/fdm/fdm_maps.py
'''
Created on Mar 11, 2013

@author: rafa
'''

import nipype.interfaces.fsl as fsl
import nipype.pipeline.engine as pe
import os
import json
from nipype.interfaces.utility import Function
import fnmatch
import numpy as np
import matplotlib.pyplot as plt
import shutil
from sfDM.vis import colormaps
from sfDM.vis.map_maker import MapMaker
import re
from nipype import config

fsl.FSLCommand.set_default_output_type('NIFTI')


#Data set up
config_file = os.environ['fdm_config']
timeline_file = os.environ['fdm_timeline']

with open(config_file, 'r') as f:
    cfg = json.load(f)
parent = cfg['parent_dir']

config.set('execution','crashdump_dir',parent)

with open(timeline_file, 'r') as f:
    tmln = json.load(f)

outputdir=parent+'/FDM/Outputs/'
try:
    threshval = float(cfg['threshold'])
except Exception, e:
    threshval = 400
    print("Could not find threshold value, using 400")

#-----------------------Begin Map Calculations--------------------------------------------------------------------#

fsl.FSLCommand.set_default_output_type('NIFTI')

scan_list = ["scan_{0:0>2d}".format(x+1) for x in range(cfg['number_of_scans'])]
print scan_list


Tumor=[outputdir+'Tumor/'+scan+'/'+(os.listdir(outputdir+'Tumor/'+scan)[0]) for scan in scan_list]
ADC=[outputdir+'ADC/'+scan+'/'+(os.listdir(outputdir+'ADC/'+scan)[0]) for scan in scan_list]
Normal=[outputdir+'Normal/'+scan+'/'+(os.listdir(outputdir+'Normal/'+scan)[0]) for scan in scan_list]


def getScans(ADClist,TumorValueList,NormalValueList, tmln):
    time1=[]
    time2=[]
    TumorValue1=[]
    TumorValue2=[]
    NormalValue1=[]
    NormalValue2=[]
    for scan in tmln['baseline']:
        time2.append(ADClist[int(scan[5:7])-1])
        time1.append(ADClist[int(scan[15:17])-1])
        NormalValue2.append(NormalValueList[int(scan[5:7])-1])
        NormalValue1.append(NormalValueList[int(scan[15:17])-1])
        TumorValue2.append(TumorValueList[int(scan[5:7])-1])
        TumorValue1.append(TumorValueList[int(scan[15:17])-1])
    for scan in tmln['serial']:
        time2.append(ADClist[int(scan[5:7])-1])
        time1.append(ADClist[int(scan[15:17])-1])
        NormalValue2.append(NormalValueList[int(scan[5:7])-1])
        NormalValue1.append(NormalValueList[int(scan[15:17])-1])
        TumorValue2.append(TumorValueList[int(scan[5:7])-1])
        TumorValue1.append(TumorValueList[int(scan[15:17])-1])

    return time1,TumorValue1,NormalValue1,time2,TumorValue2,NormalValue2

ScanSource=pe.Node(name='scansource',
                   interface=Function(input_names=['ADClist','TumorValueList','NormalValueList', 'tmln'],
                                      output_names=['time1','TumorValue1','NormalValue1',
                                                    'time2', 'TumorValue2','NormalValue2'],
                                      function=getScans))

ScanSource.inputs.tmln = tmln

outstring='-add %s '*(len(Tumor)-1)
outstring=outstring[:-1]

CreateTumorMask=pe.Node(interface=fsl.MultiImageMaths(), name='CreateTumorMask')
CreateTumorMask.inputs.in_file=Tumor[0]
CreateTumorMask.inputs.operand_files=Tumor[1:]
CreateTumorMask.inputs.op_string=outstring

ThresholdTumor=pe.Node(interface=fsl.Threshold(), name='ThresholdTumor')
ThresholdTumor.inputs.thresh=1


CreateNormalMask=pe.Node(interface=fsl.MultiImageMaths(), name='CreateNormalMask')
CreateNormalMask.inputs.in_file=Normal[0]
CreateNormalMask.inputs.operand_files=Normal[1:]
CreateNormalMask.inputs.op_string=outstring

ThresholdNormal=pe.Node(interface=fsl.Threshold(), name='ThresholdNormal')
ThresholdNormal.inputs.thresh=1

MaskTumorADC=pe.MapNode(interface=fsl.ApplyMask(), name='MaskTumorADC',iterfield=['in_file'])


ThresholdADC=pe.MapNode(interface=fsl.Threshold(), name='ThresholdADC',iterfield=['in_file'])
ThresholdADC.inputs.in_file=ADC
ThresholdADC.inputs.thresh=10


MaskNormalADC=pe.MapNode(interface=fsl.ApplyMask(), name='MaskNormalADC',iterfield=['in_file'])
MaskNormalADC.inputs.in_file=ADC


ExtractTumorValue=pe.MapNode(interface=fsl.ImageMeants(), name='ExtractTumorValues',iterfield=['in_file'])
ExtractTumorValue.inputs.show_all=True
ExtractTumorValue.inputs.transpose=True

ExtractNormalValue=pe.MapNode(interface=fsl.ImageMeants(), name='ExtractNormalValues',iterfield=['in_file'])
ExtractNormalValue.inputs.show_all=True
ExtractNormalValue.inputs.transpose=True


ExtractMaskValue=pe.Node(interface=fsl.ImageMeants(), name='ExtractMaskValues')
ExtractMaskValue.inputs.show_all = True
ExtractMaskValue.inputs.transpose = True


ImageSubtract=pe.MapNode(interface=fsl.BinaryMaths(), name='ImageSubtract', iterfield=['operand_file','in_file'])
ImageSubtract.inputs.operation='sub'


def ConfInt(value1,value2, thresh):
   # import numpy as np
   # from scipy.stats.distributions import t
    if len(value1)==len(value2):
   #     n=len(value1)
   #     dof=n-1
        #Using empirical value from literature
        pos_thresh = thresh
        neg_thresh = -thresh
        return pos_thresh, neg_thresh

    else:
        return 0

ConfidenceInterval=pe.Node(name='ConfidenceInterval',
                   interface=Function(input_names=['value1','value2', 'thresh'],
                                      output_names=['pos_thresh','neg_thresh'],
                                      function=ConfInt))
ConfidenceInterval.inputs.thresh = threshval



Maps = pe.Workflow(name='Maps')
Maps.base_dir=parent+'/FDM/'
Maps.connect([
               (ThresholdADC, MaskTumorADC, [('out_file','in_file')]),
          (CreateTumorMask, ThresholdTumor, [('out_file','in_file')]),
         (ThresholdTumor, ExtractMaskValue, [('out_file', 'in_file')]),
         (ThresholdTumor, ExtractMaskValue, [('out_file','mask')]),
             (ThresholdTumor, MaskTumorADC, [('out_file','mask_file')]),
        (CreateNormalMask, ThresholdNormal, [('out_file','in_file')]),
           (ThresholdNormal, MaskNormalADC, [('out_file','mask_file')]),
      (ThresholdNormal, ExtractNormalValue, [('out_file','mask')]),
        (MaskNormalADC, ExtractNormalValue, [('out_file','in_file')]),
        (ThresholdTumor, ExtractTumorValue, [('out_file','mask')]),
           (MaskTumorADC, ExtractTumorValue, [('out_file','in_file')]),
                  (MaskTumorADC, ScanSource, [('out_file','ADClist')]),
             (ExtractTumorValue, ScanSource, [('out_file','TumorValueList')]),
            (ExtractNormalValue, ScanSource, [('out_file','NormalValueList')]),
                 (ScanSource, ImageSubtract, [('time1','operand_file'),
                                              ('time2','in_file')]),
            (ScanSource, ConfidenceInterval, [('NormalValue1','value1'),
                                              ('NormalValue2','value2')])
])


#----------------------Begin Image creating and stats --------------------------------#

def getname(s):
    """
    Helper function to get time point from log files
    """
    try:
        Times=re.findall('_MaskTumorADC.{1,2}/',s)
        time2 = Times[0][13:-1]
        time1 = Times[1][13:-1]

        return int(time1)+1, int(time2)+1
    except ValueError:
       return "string error"

def create_image(bg,mask, overlay, X , Y , Z , savefile):
    """
    Generates 3-plane image fDM at specified coordinates
    using the map maker class given background, mask and overlay
    """
    image = MapMaker(bg)
    image.add_overlay(mask, .1, 1, colormaps.green(), alpha=0.4)
    image.add_overlay(overlay, -threshval, -3*threshval, colormaps.blue_r())
    image.add_overlay(overlay, threshval, 3*threshval, colormaps.red())
    image.save_3plane(X, Y, Z, savefile)

def create_weight_image(bg, mask, X, Y, Z, savefile):
    """
    Creates the gradient image from the weighted mask
    """
    image = MapMaker(bg)
    image.add_overlay(mask, .1, 'max',colormaps.green_r(), alpha= 0.9 )
    image.save_3plane_vertical(X,Y,Z,savefile)

def output_metrics(Time1, Time2, SN1, SN2, vectors, weights=[], se=threshval):
    """
    Calculates fDM metrics given two time points, where Time1,2 are vectors of ADC values,
    SN1,2 are string values of the time point names, vectors are PCA eigenvectors, and
    weights is a vector of weights (must match Time points)
    """
    valuedict = {}
    valuedict['timepoint'] = SN2+' - '+SN1

    weights = weights / np.max(weights)
    mn_weights = np.copy(weights)
    mn_weights[mn_weights < 1] = 0

    mx_iADC, mn_iADC, wt_iADC = 0.0, 0.0, 0.0
    mx_dADC, mn_dADC, wt_dADC = 0.0, 0.0, 0.0

    for i in np.arange(0,len(Time1)):
        if Time2[i]-Time1[i] > se:
            mx_iADC += 1
            mn_iADC += mn_weights[i]
            wt_iADC += weights[i]

        elif Time2[i]-Time1[i] < -se:
            mx_dADC += 1
            wt_dADC += weights[i]
            mn_dADC += mn_weights[i]


    mx_fiADC = mx_iADC/len(Time1)
    mx_fdADC = mx_dADC/len(Time1)
    mx_ADCratio = mx_fiADC/(0.01+mx_fdADC)


    wt_fiADC = wt_iADC/np.sum(weights)
    wt_fdADC = wt_dADC/np.sum(weights)
    wt_ADCratio = wt_fiADC / (0.01+wt_fdADC)

    mn_fiADC = mn_iADC/np.sum(mn_weights)
    mn_fdADC = mn_dADC/np.sum(mn_weights)
    mn_ADCratio = mn_fiADC / (0.01+mn_fdADC)

    mx_slope = np.arctan(vectors[0,1]/vectors[0,0]) * 180 / np.pi


    valuedict = {}
    valuedict['timepoint'] = SN2+'-'+SN1
    valuedict['mx_fiADC'] = mx_fiADC
    valuedict['mx_fdADC'] = mx_fdADC
    valuedict['mx_FDMratio'] = mx_ADCratio
    valuedict['mn_fiADC'] = mn_fiADC
    valuedict['mn_fdADC'] = mn_fdADC
    valuedict['mn_FDMratio'] = mn_ADCratio
    valuedict['wt_fiADC'] = wt_fiADC
    valuedict['wt_fdADC'] = wt_fdADC
    valuedict['wt_FDMratio'] = wt_ADCratio
    valuedict['vectors'] = vectors
    valuedict['slope'] = mx_slope
    valuedict['out_string'] = 'Timepoints,mx_fiADC,mx_fdADC,mx_fDMratio,'+\
                                         'mn_fiADC,mn_fd_ADC,mn_fDMratio,'+\
                                         'wt_fiADC,wt_fdADC,wt_fDMratio \n'+\
                                 valuedict['timepoint'] +','+\
            str(mx_fiADC) +','+ str(mx_fdADC) +','+ str(mx_ADCratio) +','+\
            str(mn_fiADC) +','+ str(mn_fdADC) +','+ str(mn_ADCratio) +','+\
            str(wt_fiADC) +','+ str(wt_fdADC) +','+ str(wt_ADCratio) +'\n'


    return valuedict


def plot_adc(results,Val1,Val2,TP1,TP2,weights,se=threshval,sfile='Time1Time2'):
    axis_lim = threshval*8
    X=np.arange(0,len(Val1))
    Scale=np.zeros((len(Val1), 3)) # matrix of RGB colors matching each voxel
    iADC = 0.0
    dADC = 0.0
    for i in X:
        if Val2[i]-Val1[i] > se:
            Scale[i] = [1,0,0]
            iADC += 1
        elif Val2[i]-Val1[i] < -se:
            Scale[i] = [0, 0, 1]
            dADC += 1
        else:
            Scale[i] = [0, 1, 0]

    weights = (weights/np.max(weights)).reshape([len(Val1),1])
    colorscale = np.hstack([Scale, weights]) # weights become the alpha channel

    fig1=plt.figure()
    ax=fig1.add_subplot(111)
    plt.scatter(Val1,Val2,c=colorscale, edgecolors='none', zorder=1)
    plt.plot(X,X,lw=3,ls='-', color='Green',alpha=0.8)
    plt.plot(X,X-se,lw=3,ls='--', color='Green',alpha=0.8)
    plt.plot(X,X+se,lw=3,ls='--', color='Green',alpha=0.8)
    plt.plot([0,-results['vectors'][0,0]*axis_lim]+Val1.mean(),
             [0,-results['vectors'][0,1]*axis_lim]+Val2.mean(),lw=3,color='Yellow')
    plt.plot([0,results['vectors'][0,0]*axis_lim]+Val1.mean(),
             [0,results['vectors'][0,1]*axis_lim]+Val2.mean(),lw=3,color='Yellow')
    fig1.set_facecolor([1,1,1])
    ax.patch.set_alpha(0.25)
    textstr = 'mx_fiADC = {:.2f} mx_fdADC = {:.2f} mx_fDMratio = {:.2f}  Slope = {:.2f}$^\circ$\n'.format(
      results['mx_fiADC'],results['mx_fdADC'],results['mx_FDMratio'],results['slope']) + \
              'mn_fiADC = {:.2f} mn_fdADC = {:.2f} mn_fDMratio = {:.2f} \n'.format(
      results['mn_fiADC'],results['mn_fdADC'],results['mn_FDMratio'])+ \
              ' wt_fiADC = {:.2f}  wt_fdADC = {:.2f}  wt_fDMratio = {:.2f}'.format(
      results['wt_fiADC'],results['wt_fdADC'],results['wt_FDMratio'])

    #ax.text(10,10, textstr, bbox={'facecolor':'white','alpha':'0.8','pad':10})
    ax.set_xlim(0,axis_lim)
    ax.set_ylim(0,axis_lim)
    ax.set_xlabel('Scan '+TP1,fontsize=14,fontweight='bold')
    ax.set_ylabel('Scan '+TP2,fontsize=14,fontweight='bold')
    #fig1.suptitle(results['timepoint'], fontsize=14, fontweight='bold')
    #ax.set_title(textstr, fontsize = 10)
    fig1.suptitle(textstr, fontsize = 10)
    plt.savefig(sfile,dpi=900, facecolor=fig1.get_facecolor(), edgecolor='w',
                orientation='landscape', bbox_inches=None, pad_inches=0.3)
    plt.close(fig1)


def get_pca(data):

    """
    compute PCA between row vectors, where each row is an observation,
    columns are samples (subjects, voxel intensity across time)
    """
    M = (data.T-np.mean(data,axis=1)).T ###subtract sample means at each observation
                                        ### observation==Time, or Slice
    [vals, vecs] = np.linalg.eig(np.cov(M)) # get eigenvales, vectors of Cov.matrix
    idx = np.argsort(vals)  #sort eigenvalues
    idx = idx[::-1]   #switch to descending order
    vecs = vecs[:,idx] #sort eigenvectors according eigenvalue order
    vals = vals[idx] #same for vals

    score = np.dot(vecs.T,M) #projection data
    return vecs, vals, score


PID=cfg['PID']
Xsl, Ysl, Zsl  = cfg['tumor_center']
charlen=len(PID)

MainDir = cfg['parent_dir']
BaseDir = MainDir+'/FDM/'
MapSearchDir = BaseDir+'Maps/ImageSubtract/mapflow/'
MapOutputDir = BaseDir+'Outputs/Maps/'
ValueSearchDir = BaseDir+'Maps/ExtractTumorValues/mapflow/'
PlotOutputDir = BaseDir+'Outputs/Plots/'
ImageOutputDir = BaseDir+'Outputs/Images/'
MaskVals= BaseDir + 'Maps/ExtractMaskValues/'
MetricsOutputDir = BaseDir+'Outputs/Metrics/'

#----------------------- Execution ---------------------------------------#
os.chdir(MainDir)
Maps.write_graph()
Maps.run()

if not os.path.isdir(MapOutputDir):
    os.makedirs(MapOutputDir)
if not os.path.isdir(PlotOutputDir):
    os.makedirs(PlotOutputDir)
if not os.path.isdir(ImageOutputDir):
    os.makedirs(ImageOutputDir)
if not os.path.isdir(MetricsOutputDir):
    os.makedirs(MetricsOutputDir)


### Copy mask into directory for easier access and viewing

for file in os.listdir(BaseDir+'Maps/CreateTumorMask'):
    if fnmatch.fnmatch(file, '*.nii'):
        shutil.copy2(BaseDir+'Maps/CreateTumorMask/'+file, MapOutputDir+PID+'_Tumor_Mask.nii')
Mask = MapOutputDir+PID+'_Tumor_Mask.nii'
BG = cfg['struct']
create_weight_image(BG, Mask, Xsl, Ysl, Zsl, ImageOutputDir+PID+'_Weights' )



#Organize and copy maps into output directory, create slice images

map_list = os.listdir(MapSearchDir)
for mapdir in map_list:
    log = MapSearchDir+mapdir+'/command.txt'
    if os.path.isfile(log):
        with open(log, 'r') as f:
            string = f.read()
        time1,time2 = getname(string)
        print 'time 2 is: scan_{0:0>2d}'.format(time2)
        print 'time 1 is: scan_{0:0>2d}'.format(time1)
        for file in os.listdir(MapSearchDir+mapdir):
            if fnmatch.fnmatch(file, '*.nii'):
                outmap = file
        print 'outmap: '+outmap
        FDM = MapOutputDir+'scan_{0:0>2d}-scan_{1:0>2d}.nii'.format(time2,time1)
        shutil.copy2(MapSearchDir+'/'+mapdir+'/'+outmap,
             FDM)
        create_image(BG, Mask, FDM, Xsl, Ysl, Zsl,\
                     ImageOutputDir+'/scan_{0:0>2d}-scan_{1:0>2d}.png'.format(time2,time1))

    else:
        print mapdir +' missing log'

def get_int(match):
    """
    Returns int value of folder name (listdir is alphabetical,not numerical )
    """
    return int(match[19:])

#### fetch text values and create plot

value_dirs = sorted(os.listdir(ValueSearchDir), key = get_int)
val_list = []
for valdir in value_dirs:
    for file in os.listdir(ValueSearchDir+valdir):
        if fnmatch.fnmatch(file, '*masked_ts.txt'):
            timepoint = file
    outval = ValueSearchDir+valdir+'/'+timepoint
    val_list.append(outval)
    print outval

for txtfile in os.listdir(MaskVals):
    if fnmatch.fnmatch(txtfile, '*_thresh_ts.txt'):
        maskvals = txtfile

weights = np.loadtxt(MaskVals+maskvals)[:,3]

#Iterate over scan time points
#calc PCA, get results dict and plot

for scan in tmln['baseline']:
    time2 = scan[5:7]
    time1 = scan[15:17]
    print 'scan {0} - scan {1}'.format(time2,time1)
    val2=np.loadtxt(val_list[int(time2)-1])[:,3]
    val1=np.loadtxt(val_list[int(time1)-1])[:,3]
    pca=get_pca(np.vstack((val1, val2)))[0]
    results=output_metrics(val1, val2, time1, time2, pca, weights)
    with open(MetricsOutputDir+'scan_{0}-scan_{1}.csv'.format(time2,time1),'wb') as f:
        f.write(results['out_string'])
    plot_adc(results,val1, val2,time1,time2,weights,
                sfile=PlotOutputDir+'scan_{0}-scan_{1}'.format(time2,time1))

for scan in tmln['serial']:
    time2 = scan[5:7]
    time1 = scan[15:17]
    print 'scan {0} - scan {1}'.format(time2,time1)
    val2=np.loadtxt(val_list[int(time2)-1])[:,3]
    val1=np.loadtxt(val_list[int(time1)-1])[:,3]
    pca=get_pca(np.vstack((val1, val2)))[0]
    results=output_metrics(val1, val2, time1, time2, pca, weights)
    with open(MetricsOutputDir+'scan_{0}-scan_{1}.csv'.format(time2,time1),'wb') as f:
        f.write(results['out_string'])
    plot_adc(results,val1, val2,time1,time2,weights,
                sfile=PlotOutputDir+'scan_{0}-scan_{1}'.format(time2,time1))

print 'Maps done'



